platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -137,6 +154,280 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static
+odp_crypto_alg_err_t aes_gcm_encrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	armv8_cipher_constants_t cc = {
+		.mode = 0
+	};
+	uint8_t *iv_ptr;
+	uint64_t nonce_length, nonce_byte_length;
+	uint64_t plaintext_length, plaintext_byte_length;
+	uint64_t aad_length, aad_byte_length;
+	int ret = 0;
+
+	/* Clear the allocated buffers */
+	memset(global->input, 0, _ARM_CRYPTO_MAX_DATA_LENGTH);
+	memset(global->nonce, 0, _ARM_CRYPTO_MAX_IV_LENGTH);
+	memset(global->aad, 0, _ARM_CRYPTO_MAX_AAD_LENGTH);
+	memset(global->digest, 0, _ARM_CRYPTO_MAX_DIGEST_LENGTH);
+	memset(global->output, 0, (_ARM_CRYPTO_MAX_DATA_LENGTH +
+	       _ARM_CRYPTO_MAX_DIGEST_LENGTH));
+
+	nonce_length = (session->p.cipher_iv.length) * 8;
+	/* IV length must be a multiple of 16 */
+	nonce_byte_length = ((nonce_length + 127) & ~127UL) >> 3;
```

Comment:
This is somewhat obfuscated. After this nonce_length is 96 bits (i.e. 12 bytes) but nonce_byte_length is 16 bytes. After peeking at the library, what you are really doing here is providing 12 byte IV to the library but making sure that the library can read 16 bytes to get the 12 byte IV.

This multiplying and rounding is actually unnecessary since this code supports only 12 byte IVs. You could simply use constants for the various lengths you need. Session creation already refuses to create the session with other IV lengths than 12.

Reply:
I did not actually think of that. I will remove these local variables and use constants for these lengths instead. 

Also, I want to clarify the difference between `nonce_length` and `nonce_byte_length` (which is also applicable for the subsequent plaintext, AAD lengths as well). As a requirement of the Armv8 crypto library, the IV/AAD/plaintext have to be padded up to the block size, which is 16 bytes. So, on the basis of the `*_length`,  I calculate the next closest multiple of 16 in `*_byte_length`. This is actually redundant for IV and AAD and I will make the change for that, but still applicable for the plaintext as its lengths vary. Probably using nonce_padded_length instead of nonce_byte_length would have been appropriate for understanding its meaning or purpose.

Reply:
> As a requirement of the Armv8 crypto library, the IV/AAD/plaintext have to be padded up to the block size, which is 16 bytes

I do not think that is an entirely correct statement. Padding, in my view, implies that additional bytes get appended in the data, making the data longer. But that is not really the requirement since library does seem to accept IV/AAD/plaintext of any length. The library may just read a bit beyond the end of the data (so those accesses better not segfault) but eventually discards the extra bytes. But it is not padding and talking about padding is creating confusion.

The GCM algorithm does require padding of the 96-bit IV to 128 bits as specified in NIST 800-38D. That padding is added by the Armv8 crypto library at the beginning of armv8_aes_gcm_set_counter() function.

> but still applicable for the plaintext as its lengths vary. 

I do not think so. The library may write a little bit beyond the end of the output data buffer, so that needs to be taken into account, but the input data buffer (plaintext in this case) does not have that requirement but just the requirement that it must be ok to read a bit past the end of the data buffer.

Reply:
Thanks for your comments, Janne. I got rid of the padding that I was adding to IV/AAD/Plaintext and ran the tests to see if library required the padding to be done in advance of calling the API. As it turns out, it works just fine without it too. I will integrate this change into the next version.